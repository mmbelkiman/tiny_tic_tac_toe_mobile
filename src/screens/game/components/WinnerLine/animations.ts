import {Animated} from 'react-native';

export const playShakeAnimation = (animation: Animated.Value) =>
  Animated.sequence([
    Animated.timing(animation, {
      toValue: 1,
      duration: 200,
      useNativeDriver: false,
    }),
    Animated.timing(animation, {
      toValue: 0,
      duration: 200,
      useNativeDriver: false,
    }),
  ]).start();

export const playSizeAnimation = (animation: Animated.Value, size: number) =>
  Animated.timing(animation, {
    toValue: size,
    duration: 80,
    useNativeDriver: false,
  }).start();

export const playOpacityAnimation = (animation: Animated.Value) =>
  Animated.timing(animation, {
    toValue: 0.6,
    duration: 90,
    useNativeDriver: false,
  }).start();
