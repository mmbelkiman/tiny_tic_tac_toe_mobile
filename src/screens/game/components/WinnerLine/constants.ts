export const STRAIGHT_LINE_SIZE_DIVISOR = 8;
export const STRAIGHT_LINE_POSITION_PADDING = 0.45;
export const DIAGONAL_LINE_SIZE_DIVISOR = 10;
export const DIAGONAL_LINE_ANGLE = 45;
export const DIAGONAL_LINE_LENGTH_SIZE_PADDING = 50;
