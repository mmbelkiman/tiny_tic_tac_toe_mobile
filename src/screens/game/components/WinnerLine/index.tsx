import React, {useEffect, useRef, useState} from 'react';
import {Animated} from 'react-native';
import {WinnerResultDirection} from '@game/common/types';
import HorizontalLine from './components/HorizontalLine';
import VerticalLine from './components/VerticalLine';
import DiagonalTopLeftLine from './components/DiagonalTopLeftLine';
import DiagonalTopRightLine from './components/DiagonalTopRightLine';
import {
  getDiagonalLineLength,
  getDiagonalLineSize,
  getStraightLinePosition,
  getStraightLineSize,
} from './utils';
import {
  playOpacityAnimation,
  playShakeAnimation,
  playSizeAnimation,
} from '@game/components/WinnerLine/animations';

interface WinnerLineProps {
  color: string | null;
  winnerResultDirection: WinnerResultDirection | null;
  boardPosition: number | null;
  squareSize: number;
  boardSize: number;
}

const WinnerLine: React.FC<WinnerLineProps> = ({
  color,
  boardPosition,
  squareSize,
  winnerResultDirection,
  boardSize,
}) => {
  const [lineSize, setLineSize] = useState(0);
  const opacityAnimation = useRef(new Animated.Value(0)).current;
  const sizeAnimation = useRef(new Animated.Value(0)).current;
  const shakeAnimation = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    //Reset animations when board is reset
    if (color === null) {
      opacityAnimation.resetAnimation();
      sizeAnimation.resetAnimation();
      shakeAnimation.resetAnimation();
      setLineSize(0);
    }
  }, [sizeAnimation, opacityAnimation, shakeAnimation, color]);

  useEffect(() => {
    //Set the line size
    if (winnerResultDirection === null) {
      return;
    }
    if (
      winnerResultDirection === 'horizontal' ||
      winnerResultDirection === 'vertical'
    ) {
      setLineSize(getStraightLineSize(squareSize));
    }
    setLineSize(getDiagonalLineSize(squareSize));
  }, [squareSize, winnerResultDirection]);

  useEffect(() => {
    //Run animations when has a winner
    if (color !== null) {
      playOpacityAnimation(opacityAnimation);
      playSizeAnimation(sizeAnimation, lineSize);
      playShakeAnimation(shakeAnimation);
    }
  }, [sizeAnimation, opacityAnimation, shakeAnimation, lineSize, color]);

  if (color === null || boardPosition === null) {
    return null;
  }

  if (winnerResultDirection === 'horizontal') {
    return (
      <HorizontalLine
        testID={boardPosition}
        color={color}
        positionY={getStraightLinePosition(squareSize, boardPosition)}
        opacityAnimation={opacityAnimation}
        shakeAnimation={shakeAnimation}
        heightAnimation={sizeAnimation}
      />
    );
  }

  if (winnerResultDirection === 'vertical') {
    return (
      <VerticalLine
        testID={boardPosition}
        width={lineSize}
        color={color}
        positionX={getStraightLinePosition(squareSize, boardPosition)}
        opacityAnimation={opacityAnimation}
        shakeAnimation={shakeAnimation}
        widthAnimation={sizeAnimation}
      />
    );
  }

  if (winnerResultDirection === 'diagonalTopLeft') {
    return (
      <DiagonalTopLeftLine
        testID={boardPosition}
        height={lineSize}
        color={color}
        length={getDiagonalLineLength(boardSize, squareSize)}
        opacityAnimation={opacityAnimation}
        shakeAnimation={shakeAnimation}
        heightAnimation={sizeAnimation}
      />
    );
  }
  return (
    <DiagonalTopRightLine
      testID={boardPosition}
      height={lineSize}
      color={color}
      length={getDiagonalLineLength(boardSize, squareSize)}
      opacityAnimation={opacityAnimation}
      shakeAnimation={shakeAnimation}
      heightAnimation={sizeAnimation}
    />
  );
};

export default WinnerLine;
