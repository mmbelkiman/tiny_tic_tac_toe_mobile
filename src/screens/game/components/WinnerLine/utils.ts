import {
  DIAGONAL_LINE_ANGLE,
  DIAGONAL_LINE_LENGTH_SIZE_PADDING,
  DIAGONAL_LINE_SIZE_DIVISOR,
  STRAIGHT_LINE_POSITION_PADDING,
  STRAIGHT_LINE_SIZE_DIVISOR,
} from './constants';

export const getStraightLineSize = (squareSize: number) =>
  squareSize / STRAIGHT_LINE_SIZE_DIVISOR;
export const getStraightLinePosition = (
  squareSize: number,
  boardPosition: number,
) => (boardPosition + STRAIGHT_LINE_POSITION_PADDING) * squareSize;
export const getDiagonalLineSize = (squareSize: number) =>
  squareSize / DIAGONAL_LINE_SIZE_DIVISOR;

export const getDiagonalLineLength = (
  boardSize: number,
  squareSize: number,
) => {
  const squareDiagonalLength = Math.sqrt(squareSize * squareSize * 2);
  const lineLength =
    boardSize * Math.cos(DIAGONAL_LINE_ANGLE) * squareDiagonalLength;

  //multiply by 2 to cover the whole board
  return lineLength * 2 - DIAGONAL_LINE_LENGTH_SIZE_PADDING;
};
