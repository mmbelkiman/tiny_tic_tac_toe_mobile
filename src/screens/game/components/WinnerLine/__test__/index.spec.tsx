import React from 'react';
import renderer from 'react-test-renderer';
import WinnerLine from '../index';

// Mock inside components, just check if the correct one is rendered
jest.mock('../../HorizontalLine', () => 'HorizontalLine');
jest.mock('../../VerticalLine', () => 'VerticalLine');
jest.mock('../../DiagonalTopLeftLine', () => 'DiagonalTopLeftLine');
jest.mock('../../DiagonalTopRightLine', () => 'DiagonalTopRightLine');

describe('WinnerLine Component', () => {
  it('should match snapshot when no winner', () => {
    const tree = renderer
      .create(
        <WinnerLine
          winnerMark=""
          boardPosition={null}
          squareSize={100}
          winnerResultDirection={null}
          boardSize={10}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match the horizontal line snapshot', () => {
    const tree = renderer
      .create(
        <WinnerLine
          winnerMark="X"
          boardPosition={3}
          squareSize={100}
          winnerResultDirection="horizontal"
          boardSize={10}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match the vertical line snapshot', () => {
    const tree = renderer
      .create(
        <WinnerLine
          winnerMark="O"
          boardPosition={5}
          squareSize={100}
          winnerResultDirection="vertical"
          boardSize={8}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match the diagonal top left line snapshot', () => {
    const tree = renderer
      .create(
        <WinnerLine
          winnerMark="X"
          boardPosition={null}
          squareSize={100}
          winnerResultDirection="diagonalTopLeft"
          boardSize={5}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match the diagonal top right line snapshot', () => {
    const tree = renderer
      .create(
        <WinnerLine
          winnerMark="O"
          boardPosition={2}
          squareSize={100}
          winnerResultDirection="diagonalTopRight"
          boardSize={6}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
