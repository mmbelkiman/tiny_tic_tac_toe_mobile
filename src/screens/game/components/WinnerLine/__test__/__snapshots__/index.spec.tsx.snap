// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WinnerLine Component should match snapshot when no winner 1`] = `null`;

exports[`WinnerLine Component should match the diagonal top left line snapshot 1`] = `
<DiagonalTopLeftLine
  color="red"
  height={0}
  heightAnimation={0}
  length={692.9174811988408}
  opacityAnimation={0}
  shakeAnimation={0}
/>
`;

exports[`WinnerLine Component should match the diagonal top right line snapshot 1`] = `
<DiagonalTopRightLine
  color="red"
  height={10}
  heightAnimation={0}
  length={841.5009774386089}
  opacityAnimation={0}
  shakeAnimation={0}
/>
`;

exports[`WinnerLine Component should match the horizontal line snapshot 1`] = `
<HorizontalLine
  color="red"
  heightAnimation={0}
  opacityAnimation={0}
  positionY={345}
  shakeAnimation={0}
/>
`;

exports[`WinnerLine Component should match the vertical line snapshot 1`] = `
<VerticalLine
  color="red"
  opacityAnimation={0}
  positionX={545}
  shakeAnimation={0}
  width={10}
  widthAnimation={0}
/>
`;
